### ISSUE TITLE
Non-root commands are not logged if "file" output is used and target file is not writeable by non-root users

### ISSUE THREAD

**Issue by GOPiGal** [a2o/snoopy#133]
Remote execution of command via SSH is logged only as execution of sshd with  -D -R arguments.
It is reported in #95 and ~~incorrectly~~ correctly marked as incorrect report :-)

Architecture: _x86_64_
Linux distribution: _Ubuntu 16.04.3 Xenial Xerus_
Snoopy version: _2.4.6_
Snoopy config file was used: _yes (see below)_
Snoopy installed with _snoopy-install.sh_

_/etc/snoopy.ini_ on the **dest.host**

```
[snoopy]
message_format = "timestamp=%{datetime} host=%{hostname} uid=%{uid} uname=%{login} euid=%{euid} euname=%{eusername} pid=%{pid} ppid=%{ppid} tty=%{tty} cwd=%{cwd} cmdline=%{cmdline}"
output = file:/var/log/snoopy.log
```

>**Quick explanation on sanitization of following transcripts** 
>
> - rem.host below is ip of remote host used to execute command via ssh on the dest.host
> - dest.host is IP address of machine with installed snoopy 

After the reboot following commands were executed:

> dest.host:# aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
> **_rem.host_**:$ ssh pg@dest.host parted
> dest.host:# aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa

Expected results were logs  from attempted execution of 

- (twice) unknown commands **aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa** by **root** 
- know command **parted** by user **pg**

Result below:

_/var/log/snoopy.log_
```
timestamp=2018-04-11T16:32:08+0200 host=dest.host uid=0 uname=pg euid=0 euname=root pid=3248 ppid=3247 tty=/dev/pts/4 cwd=/root cmdline=/usr/lib/command-not-found -- aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
timestamp=2018-04-11T16:32:14+0200 host=dest.host uid=0 uname=(unknown) euid=0 euname=root pid=3250 ppid=923 tty=(none) cwd=/ cmdline=/usr/sbin/sshd -D -R
timestamp=2018-04-11T16:32:21+0200 host=dest.host uid=0 uname=pg euid=0 euname=root pid=3280 ppid=3279 tty=/dev/pts/4 cwd=/root cmdline=/usr/lib/command-not-found -- aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
timestamp=2018-04-11T16:32:34+0200 host=dest.host uid=0 uname=pg euid=0 euname=root pid=3283 ppid=2505 tty=/dev/pts/4 cwd=/root cmdline=cat /var/log/snoopy.log

```
which summarize into (second line incorrect):

- (twice) unknown commands **aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa** by **root** 
- _**sshd -D -R** by user **root**_

Below is the auth.log confirming ssh session.

_/var/log/auth.log_
```
Apr 11 16:32:17 dest.host sshd[3250]: Accepted password for pg from rem.host port 54710 ssh2
Apr 11 16:32:17 dest.host sshd[3250]: pam_unix(sshd:session): session opened for user pg by (uid=0)
Apr 11 16:32:17 dest.host systemd-logind[798]: New session 12 of user pg.
Apr 11 16:32:17 dest.host sshd[3275]: Received disconnect from rem.host port 54710:11: disconnected by user
Apr 11 16:32:17 dest.host sshd[3275]: Disconnected from rem.host port 54710
Apr 11 16:32:17 dest.host sshd[3250]: pam_unix(sshd:session): session closed for user pg
Apr 11 16:32:17 dest.host systemd-logind[798]: Removed session 12.
```

**Comment by bostjan** [comment]
What is the purpose of not using the syslog?

From what I can tell (and I am guessing your /var/log/snoopy.log file is owned by root and of 644 permissions) you are trying to write as a non-root user to root-owned file.

Is that the case (remember that snoopy runs in the context of each process, inheriting its UID/GID)?

**Comment by bostjan** [comment]
Actually, here:
https://github.com/a2o/snoopy/blob/master/etc/snoopy.ini.in#L111

Quote:
```
; - file     ; (available=@enable_output_file@) Write directly to file. (NOTICE: Make sure file has proper permissions set for non-root users.)
```

**Comment by GOPiGal** [comment]
Hello,

I salute you, Sir and I enroll to school to learn reading ;-(
The issue I had was related to FS permissions, after adjusting them it works fine.

Thank you for your support & Regards

**Comment by bostjan** [comment]
Excellent :)

**Point 1:**
@GOPiGal, if you have any idea how to document this better, feel free to submit a pull request with documentation updates.

**Point 2:**
Why are you not using syslog but instead opting for direct to-file logging? I would like to understand the use case behind this, hence my asking. Thank you for letting me know.

**Comment by GOPiGal** [comment]
Point 1: The only thing I can think of is to put the information in the example conf file as commented line next the one describing how to set it up

Point 2:  I am pulling it into Splunk and if I have it in a separate file it will easy parsing events. (Not by far but...). However chmod 6666 on the log is a game changer - users would easily access the log which is not necessarily my intent so I plan now to revert to syslog.

**Comment by bostjan** [comment]
Like this?
```diff
diff --git a/etc/snoopy.ini.in b/etc/snoopy.ini.in
index e3e09cd..90cc383 100644
--- a/etc/snoopy.ini.in
+++ b/etc/snoopy.ini.in
@@ -131,7 +131,7 @@
 ; Example:
 ;output = console
 ;output = devlog
-;output = file:/var/log/snoopy.log
+;output = file:/var/log/snoopy.log   ; NOTICE: Needs appropriate file permissions. Using "root:root:0666" will only allow logging root's commands.
 ;output = socket:/var/run/socket-for-snoopy.sock
 
 
```

**Comment by GOPiGal** [comment]
:-) Or like this:
-;output = file:/var/log/snoopy.log
+;output = file:/var/log/snoopy.log   ; NOTICE: Needs appropriate file permissions for all users that are to be snoopied. Using "root:root:0666" will only allow logging root's commands.

**Comment by bostjan** [comment]
Looks good. Please create pull request for this :)

**Comment by bostjan** [comment]
Done in commit f757b733 included in version 2.4.5.
